addons/sms/controllers/main.py

```diff
@@ -0,0 +1,51 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import logging
+import re
+from operator import itemgetter
+
+from odoo import http
+from odoo.exceptions import UserError
+
+_logger = logging.getLogger(__name__)
+
+
+class SmsController(http.Controller):
+
+    @http.route('/sms/status', type='json', auth='public')
+    def update_sms_status(self, message_statuses):
+        """Receive a batch of delivery reports from IAP
+
+        :param message_statuses:
+            [
+                {
+                    'sms_status': status0,
+                    'uuids': [uuid00, uuid01, ...],
+                }, {
+                    'sms_status': status1,
+                    'uuids': [uuid10, uuid11, ...],
+                },
+                ...
+            ]
+        """
+        SmsSms_sudo = http.request.env['sms.sms'].sudo()
+        SmsTracker_sudo = http.request.env['sms.tracker'].sudo()
+        all_uuids = []
+        for uuids, iap_status in map(itemgetter('uuids', 'sms_status'), message_statuses):
+            # Basic checks to avoid unnecessary queries and allow debugging.
+            if (
+                not uuids
+                or not iap_status
+                or not re.match(r'^\w+$', iap_status)
+                or any(not re.match(r'^[0-9a-f]{32}$', uuid) for uuid in uuids)
+            ):
+                _logger.warning('Received ill-formatted SMS delivery report event: \n%s', message_statuses)
+                raise UserError("Bad parameters")
```

Comment:
As it is a public controller, shouldn't we raise BadRequest ?

Reply:
Actually, inside a JsonRPC controller, we are not really doing http but rather jsonrpc. So my money is on `UserError`. The error handling for jsonrpc errors is going to treat both errors the same way: wrap then in a jsonrpc `{"error": {}}` message and serialize the error in json.

So you'll have either

```json
{"jsonrpc": "2.0", "id": ..., "error": {
  "name": "werkzeug.exceptions.BadRequest",
  ...
}
```
or in the other case
```json
{"jsonrpc": "2.0", "id": ..., "error": {
  "name": "odoo.exceptions.UserError",
  ...
}
```

I think in case of an `UserError` you'll get a warning inside the logs, but not for a `BadRequest`